---
title: "Alpha Diversity:</br>Summarizing Microbial Communities"
subtitle: "üìöEPID 674üìö"  
author: "Brendan J. Kelly, MD, MS"
date: 'Updated: `r format(Sys.Date(), "%d %B %Y")`'
output:
  xaringan::moon_reader:
    self_contained: TRUE
    lib_dir: libs
    css: xaringan-themer-plus.css
    nature:
      ratio: 16:9
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---
background-image: url(svg/candyjar.svg)
background-size: 500px
background-position: 99% 50%
class: middle, inverse

```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
```

```{r xaringan-themer, include=FALSE, warning=FALSE, eval=FALSE}
library(xaringanthemer)
style_duo_accent(
  primary_color = "#011F5B",
  secondary_color = "#990000",
  inverse_header_color = "#FFFFFF",
  header_font_google = google_font("Roboto Condensed"),
  text_font_google = google_font("Roboto"),
  code_font_google = google_font("Roboto Mono"),
  # padding = "64px 64px 64px 64px"
  # base_font_size = "24px",
  # text_font_base = "1rem",
  # header_h1_font_size = "2.75rem",
  # header_h2_font_size = "2rem",
  # header_h3_font_size = "1.25rem",
)
```

.pad-left[

### R/tidyverse recap

### Summarizing OTU tables

### Alpha diversity

### Rarefaction vs rarefying

### R's vegan package

]



---
background-image: url(svg/candyjar.svg)
background-size: 500px
background-position: 99% 50%
class: center, middle, inverse

# R/tidyverse recap



---

# R: Making New Variables

.pad-left[
- `tidyverse` to create new variables

- Make new variables & keep old variables:

    - `mutate()` function
    
    - must return same length as input, or length 1 (wll be recycled)

- Make a new variable & lose old variables:

    - `summarise()` function (often follows `group_by()`)
    
    - simplest use case: return length of 1

]



---

# R: Making New Variables

.pull-left[

```{r package-install, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}
# install tidyverse once 
install.packages('tidyverse')

# load tidyverse functions each time you use
library(tidyverse)

```


]

.pull-right[

- Code at left installs and loads the `tidyverse` package

- The `tidyverse` package includes a set of other packages that permit streamlined data processing

- See Hadley Wickham's _R For Data Science_: [https://r4ds.had.co.nz/](https://r4ds.had.co.nz/)


]



---

# `mutate()`

.pull-left[

```{r read-otu, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}
# make sure tidyverse loaded
library(tidyverse)

# load (trimmed) HMP V1-V3 OTU table 
otu <- read_csv(
  file = 
    "./data/HMP_OTU_table_longformat.csv.gz",
)

otu # show what you've read

```


]

.pull-right[

```{r read-otu-out, ref.label="read-otu", echo=FALSE, message=FALSE, warning=FALSE}

```


]




---

# `mutate()`

.pull-left[

```{r mutate1, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

otu %>%
  mutate(
    HMPbodysubsite =
           gsub("_"," ",HMPbodysubsite)
    )



```


]

.pull-right[

```{r mutate1-out, ref.label="mutate1", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# `mutate()`

.pull-left[

```{r mutate2, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

otu %>%
  mutate(
    HMPbodysubsite =
           gsub("_"," ",HMPbodysubsite),
    log_reads = log10(read_count)
    )



```


]

.pull-right[

```{r mutate2-out, ref.label="mutate2", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# `summarise()`

.pull-left[

```{r summarise1, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

otu %>%
  group_by(HMPbodysubsite) %>%
  summarise(
    mean_reads = mean(read_count, na.rm = TRUE)) %>%
  ungroup()


```


]

.pull-right[

```{r summarise1-out, ref.label="summarise1", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# `summarise()`

.pull-left[

```{r summarise2, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

otu %>%
  group_by(HMPbodysubsite) %>%
  summarise(
    mean_reads = mean(read_count, na.rm = TRUE),
    sum_reads = sum(read_count, na.rm = TRUE)
    ) %>%
  ungroup()


```


]

.pull-right[

```{r summarise2-out, ref.label="summarise2", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---
background-image: url(svg/candyjar.svg)
background-size: 500px
background-position: 99% 50%
class: center, middle, inverse

# Summarizing OTU tables



---

# High Dimensional Microbiome Data

.pad-left[

- Typical OTU table orientation in microbiome studies:
    - 43140 rows
    - 32 columns

]

.pull-left[

```{r otu1, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

# TYPICAL OTU TABLE ORIENTATION IN MICROBIOME STUDIES

otu %>%
  reshape2::acast(otu_id ~ specimen_id,
                  # rows = otu_id, columns = specimen_id
                  value.var = "read_count") %>%
     .[1:10,1:5]

# 43140 ROWS & 32 COLUMNS


```


]

.pull-right[

```{r otu1-out, ref.label="otu1", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# High Dimensional Microbiome Data

.pad-left[

- Typical species table orientation in ecology studies:
    - 32 rows
    - 43140 columns

]

.pull-left[

```{r otu2, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

# TYPICAL SPECIES TABLE ORIENTATION IN ECOLOGY STUDIES

otu %>%
  reshape2::acast(specimen_id ~ otu_id,
                  # rows = specimen_id, columns = otu_id
                  value.var = "read_count") %>%
   .[1:10,1:5]

# 32 ROWS & 43140 COLUMNS


```


]

.pull-right[

```{r otu2-out, ref.label="otu2", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# High Dimensional Microbiome Data

.pad-left[

- How to deal with high-dimensional microbiome data?

- __Descriptive (e.g., heatmaps and stacked barplots)__
    
- Test a priori hypotheses regarding specific OTUs/taxa

- Reduce dimensions:

    - single summary statistic (alpha diversity)
    
    - pairwise distances (beta diversity) with PCoA or PERMANOVA
    
    - community types (mixture modeling)


]




---

# Descriptive Plots

.pad-left[

- Visualization of OTU table:

    - typically present counts as a proportion of sample total
    
    - choice of sample order can highlight group differences

- Limitations:

    - cannot depict full list of OTUs
    
    - space dictates taxonomic level presented

]



---
background-image: url(img/hmp_heatmap.png)
background-size: contain



---
background-image: url(img/hmp_barplot.png)
background-size: contain



---

# High Dimensional Microbiome Data

.pad-left[

- How to deal with high-dimensional microbiome data?

- Descriptive (e.g., heatmaps and stacked barplots)
    
- __Test a priori hypotheses regarding specific OTUs/taxa__

- Reduce dimensions:

    - single summary statistic (alpha diversity)
    
    - pairwise distances (beta diversity) with PCoA or PERMANOVA
    
    - community types (mixture modeling)


]



---

# Single-Taxon Hypotheses

.pad-left[

- You suspect Bacteroides has a relationship with outcome of interest...

    - _Bacteroides_ (genus)?
    
    - _Bacteroidaceae_ (family)?
    
    - _Bacteroidales_ (order)?
    
    - _Bacteroidetes_ (class)?

- Hypotheses focusing on specific taxa often fail to account for possibility of selection bias from culture.


]



---

# High Dimensional Microbiome Data

.pad-left[

- How to deal with high-dimensional microbiome data?

- Descriptive (e.g., heatmaps and stacked barplots)
    
- Test a priori hypotheses regarding specific OTUs/taxa

- __Reduce dimensions:__

    - __single summary statistic (alpha diversity)__
    
    - pairwise distances (beta diversity) with PCoA or PERMANOVA
    
    - community types (mixture modeling)


]



---
background-image: url(svg/candyjar.svg)
background-size: 500px
background-position: 99% 50%
class: center, middle, inverse

# Alpha Diversity



---

# Alpha Diversity

.pad-left[

- One solution to the p >> n problem:  

    - capture entire community with a single numerical value
    
    - richness: what's there?
    
    - evenness: how are members distributed?
    
    - diversity: richness + evenness


]



---

# Alpha Diversity

.pad-left[

- Many alpha diversity metrics (weight richness/evenness):

    - species number, Chao1 (singletons & doubletons)
    
    - Shannon diversity: 
    
    $$H' = - \sum{ p_{i} * \log_{b}{(p_{i})} }$$
    
    (note: typically natural log or base 2 are used)


]



---

# Aside on Information Theory

.pad-left[

- Shannon diversity:

    $$H' = - \sum{ p_{i} * \log_{b}{(p_{i})} }$$
    
- Claude Shannon & information entropy:

    $$H(p) = - \sum{ p_{i} * \log_{b}{(p_{i})} }$$
    
"The uncertainty contained in a probability distribution is the average log-probability of an event." (McElreath _Statistical Rethinking, 2nd_ 2020)

]



---

# Alpha Diversity: Chao1

.pad-left[

- The Chao1 index estimates the total species __richness__:

    $$S_{Chao1} = S_{obs} + \frac{n_{1}^2}{2*n_{2}}$$
    
(S = number of species; n1 = number singletons; n2 = number doubletons)
    
- Chao1 is particularly useful for data sets skewed toward the low-abundance classes, as is likely to be the case with microbes.


]



---
background-image: url(img/hmp_shannon.png)
background-size: contain



---
background-image: url(svg/candyjar.svg)
background-size: 500px
background-position: 99% 50%
class: center, middle, inverse

# Rarefaction vs Rarefying



---
background-image: url(img/counting_uncountable.png)
background-size: contain



---

# Microbes Too Diverse to Count?

.pad-left[

- ‚ÄúIn any community, the number of types of organisms observed increases with sampling effort until all types are observed.‚Äù

- ‚ÄúThe relationship between the number of types observed and sampling effort gives information about the total diversity of the sampled community.‚Äù

- ‚ÄúPattern can be visualized by plotting an accumulation or a rank-abundance curve.‚Äù


]



---

# Microbes Too Diverse to Count?

.pad-left[

- ‚ÄúAn accumulation curve is a plot of the cumulative number of types observed versus sampling effort.‚Äù 

- ‚ÄúBecause all communities contain a finite number of species, if the surveyors continued to sample, the curves would eventually reach an asymptote at the actual community richness (number of types).‚Äù

- ‚ÄúThe curves contain information about how well the communities have been sampled (i.e., what fraction of the species in the community have been detected).‚Äù


]



---
background-image: url(img/accumulation_curves.png)
background-size: contain



---

# Microbes Too Diverse to Count?

.pad-left[

- ‚ÄúThe more concave-downward the curve, the better sampled the community.‚Äù

- ‚ÄúThe idea that microbial diversity cannot be estimated comes from the fact that many microbial accumulation curves are linear or close to linear because of high diversity, small sample sizes, or both.‚Äù

- ‚ÄúUltimately, microbes‚Äîlike tropical insects‚Äîare too diverse to count exhaustively.‚Äù


]



---
background-image: url(img/rarefaction.png)
background-size: contain



---

# Rarefaction & Rarefying

.pad-left[

- ‚ÄúRarefaction was originally introduced as a method of estimating species richness (i.e., an alternative to parametric or nonparametric estimators like Chao1)"

- ‚ÄúRarefaction compares observed richness among sites, treatments, or habitats that have been unequally sampled. A rarefied curve results from averaging randomizations of the observed accumulation curve‚Ä¶. The variance around the repeated randomizations allows one to compare the observed richness among samples.‚Äù



]



---
background-image: url(img/inadmissable.png)
background-size: contain




---

# Rarefaction & Rarefying

.pad-left[

- ‚ÄúMicrobiome analysis workflows often begin with an ad hoc library size normalization by random subsampling without replacement, or so-called rarefying‚Ä¶. There is confusion in the literature regarding terminology, and sometimes this normalization approach is conflated with a non-parametric resampling technique ‚Äî called rarefaction, or individual-based taxon resampling curves ‚Äî that can be justified for coverage analysis or species richness estimation in some settings.‚Äù



]




---

# Rarefaction & Rarefying

.pad-left[

- ‚ÄúHere we emphasize the distinction between taxon re-sampling curves and normalization by strictly adhering to the terms rarefying or rarefied counts when referring to the normalization procedure, respecting the original definition for rarefaction.‚Äù

- "Rarefying defined by the following steps: (1) Select a minimum library size... (2) Discard libraries (microbiome samples) that have fewer reads than [minimum library size]... (3) Subsample the remaining libraries without replacement such that they all have [same] size"

- ‚ÄúRarefying is now an exceedingly common precursor to microbiome multivariate workflows that seek to relate sample covariates to sample-wise distance matrices.‚Äù


]




---

# Rarefaction & Rarefying

.pad-left[

- ‚ÄúDespite its current popularity in microbiome analyses __rarefying biological count data is statistically inadmissible__ because it requires the omission of available valid data. This holds even if repeated rarefying trials are compared for stability as previously suggested.‚Äù


]




---
background-image: url(svg/candyjar.svg)
background-size: 500px
background-position: 99% 50%
class: center, middle, inverse

# R's vegan package



---

# `vegan::diversity()`

.pull-left[

```{r diversity1, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

library(vegan)

diversity(x = c(100,0,5,10), index = "shannon", base = 2)

diversity(x = c(5,5,5,5), index = "shannon", base = 2)



```


]

.pull-right[

```{r diversity1-out, ref.label="diversity1", echo=FALSE, message=FALSE, warning=FALSE}

```


]






---

# `vegan::rrarefy()`

.pull-left[

```{r rarefy1, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

library(vegan)

rrarefy(x = c(100,0,5,10), sample = 10)

rrarefy(x = c(1421,170,205,3607), sample = 1000)



```


]

.pull-right[

```{r rarefy1-out, ref.label="rarefy1", echo=FALSE, message=FALSE, warning=FALSE}

```


]




---
class: center, middle, inverse
background-image: url(svg/conjugation.svg)
background-size: 500px
background-position: 50% 50%

# Questions?
### Post to the discussion board!


---
background-image: url(svg/bacteria.svg)
background-size: 100px
background-position: 98% 90%
class: center, middle

# Thank you!
#### Slides available: [github.com/bjklab](https://github.com/bjklab/EPID674_002_sequences-to-counts.git)
#### [brendank@pennmedicine.upenn.edu](brendank@pennmedicine.upenn.edu)




